x/staking/keeper/msg_server.go

```diff
@@ -719,3 +707,37 @@ func (k msgServer) RotateConsPubKey(ctx context.Context, msg *types.MsgRotateCon
 
 	return res, nil
 }
+
+// consKeyAlreadyUsed returns true if the consensus public key is already used, both in the OldToNewConsAddrMap or in the
+// current block (RotationHistory).
+func (k msgServer) checkConsKeyAlreadyUsed(ctx context.Context, consPubKey cryptotypes.PubKey) error {
+	rotatedTo, err := k.ConsAddrToValidatorIdentifierMap.Get(ctx, consPubKey.Address())
+	if err != nil && !errors.Is(err, collections.ErrNotFound) {
+		return err
+	}
+
+	if rotatedTo != nil {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidAddress,
+			"the new public key is already present in rotation history, please try with a different one")
+	}
+
+	// check in the current block
+	rotationHistory, err := k.GetBlockConsPubKeyRotationHistory(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, rotation := range rotationHistory {
+		if bytes.Equal(rotation.NewConsPubkey.GetCachedValue().(cryptotypes.PubKey).Address(), consPubKey.Address()) {
```

Comment:
Maybe here check that `GetCachedValue() != nil`. If the the `AnyUnpack` ran ok it shouldn't be nil .. but just to be sure.

Reply:
should I try to unpack or just return an error?

Reply:
I think it should return an error, bc if that happens the `NewConsPubkey` might not being created with any.NewWithCache